mlflow/store/artifact/databricks_artifact_repo.py

```diff
@@ -231,6 +240,85 @@ def _azure_upload_file(self, credentials, local_file, artifact_path):
         except Exception as err:
             raise MlflowException(err)
 
+    def _azure_adls_gen2_upload_file(self, credentials, local_file, artifact_path):
+        """
+        Uploads a file to a given Azure storage location using the ADLS gen2 API.
+        """
+        try:
+            headers = self._extract_headers_from_credentials(credentials.headers)
+
+            # try to create the file
+            try:
+                put_adls_file_creation(credentials.signed_uri, headers=headers)
+            except requests.HTTPError as e:
+                if e.response.status_code in [401, 403]:
+                    _logger.info(
+                        "Failed to authorize ADLS file creation request, possibly due "
+                        "to credential expiration. Refreshing credentials and trying again..."
+                    )
+                    credential_info = self._get_write_credential_infos(
+                        run_id=self.run_id, paths=[artifact_path]
+                    )[0]
+                    put_adls_file_creation(credential_info.signed_uri, headers=headers)
```

Comment:
Under what circumstances does this special handling of refreshing credentials will actually work? or why does it fail in the first place that we only need to retry exactly once to resolve?

Reply:
In my understanding, presigned urls will expire after 300s or so (maybe 900s, I don't remember the exact number), if there are many chunks and the uploading takes longer than that, the `put` will fail. In this case, getting a new presigned url for write will resolve this issue.

Reply:
OK. This makes sense to me.

Reply:
@kriscon-db Can we try to abstract away this try-except block with an explicit aim to retry with refreshed credentials? It's very repetitive w/o clear indication of its functionality. Something like `def retry_with_credential_refresh(fn, credentials):`, or make it a decorator if you want.

Reply:
Yep.  Let me push a fix.

Reply:
Please re-review this section.